### DISCUSSION TITLE
each while throws 'expected block, closure or record'

### DISCUSSION THREAD

**Discussion by aserowy** [nushell/nushell#9885] [Q&A] [answered]
Heho,

after 0.83 my written scripts to change workspaces in sway are faulting and i dont get why. The script creates a list with workspace names and resolves the 'next' workspace name to navigate to.

```
#!/usr/bin/env nu

def main [direction: string] {

    let output = (swaymsg -t get_outputs 
        | from json 
        | where focused == true
        | get name
        | first)

    let workspaces = (swaymsg -t get_workspaces -r 
        | from json 
        | where output == $output
        | select name focused
        | if $direction == 'prev' {
            sort-by name --natural } else {
            sort-by name --natural --reverse })

    let workspaces = ($workspaces
        | prepend { name: ($workspaces | last | get name) focused: false })

    let next = ($workspaces 
        | each while { |workspace| if $workspace.focused != true { $workspace.name }}
        | last)

    run-external --redirect-stderr 'swaymsg' workspace $next
}
```

The `each while` in line 23 throws the following. I NEED HELP to understand the error :D

```sh
Error: nu::parser::parse_mismatch

  × Parse mismatch during operation.
    ╭─[/home/serowy/src/dots/shell/sway/sway/src/scripts/focus_workspace.nu:22:1]
 22 │     let next = ($workspaces
 23 │         | each while { |workspace| if $workspace.focused != true { $workspace.name } }
    ·                ──┬──
    ·                  ╰── expected block, closure or record
 24 │         | last)
    ╰────
```

Kind regards
Alexander

**Comment by fdncred** [comment]
I don't have/use swaymsg. Is there a way to reproduce this without it? I suspect this is a problem with `each while`'s input_output_types. Maybe if you can print out what `$workspaces` looks like, we can just use that data to pipe into it and see if we can fix it. Right now it looks like `each while` will only take a `list<any>` or a `table` as input. I'm guessing you have a record.
![image](https://github.com/nushell/nushell/assets/343840/8175936d-6be5-43e2-8382-0a64aeb6494e)

You could test it out if you're willing to try and compile nushell with this change in each_while.rs. Just replace the old signature with this new one. I'm hoping that fixes it.
```rust
    fn signature(&self) -> nu_protocol::Signature {
        Signature::build(self.name())
            .input_output_types(vec![
                (
                    Type::List(Box::new(Type::Any)),
                    Type::List(Box::new(Type::Any)),
                ),
                (Type::Table(vec![]), Type::List(Box::new(Type::Any))),
                (Type::Record(vec![]), Type::List(Box::new(Type::Any))),
            ])
            .required(
                "closure",
                SyntaxShape::Closure(Some(vec![SyntaxShape::Any, SyntaxShape::Int])),
                "the closure to run",
            )
            .allow_variants_without_examples(true)
            .category(Category::Filters)
    }
```

**Reply by aserowy** [reply]
Heho,

ty for taking the time.

I added a print and `$workspaces | describe` and got the following output:

```
╭───┬────────┬─────────╮
│ # │  name  │ focused │
├───┼────────┼─────────┤
│ 0 │ social │ false   │
│ 1 │ d4     │ false   │
│ 2 │ dots   │ true    │
│ 3 │ gaming │ false   │
│ 4 │ grow   │ false   │
│ 5 │ social │ false   │
╰───┴────────┴─────────╯

table<name: string, focused: bool>
```

Thus, the input type should be valid. Any idea, why it is failing? Btw. the error gets thrown in the 'compile' step. I had to comment out the lines to get the script running.

Regards
Alexander

**Reply by fdncred** [reply]
I can't reproduce it with this
```nushell
❯ let workspaces = [[name focused]; [social false] [d4 false] [dots true] [gaming false] [grow false] [social false]]
❯ $workspaces | each while { |workspace| if $workspace.focused != true { $workspace.name }}
╭───┬────────╮
│ 0 │ social │
│ 1 │ d4     │
╰───┴────────╯
❯ $workspaces | each while { |workspace| if $workspace.focused != true { $workspace.name }} | last
d4
```
What version of nushell are you running? I'm on 0.83.2 commit 94bec72

**Reply by aserowy** [reply]
0.83.1

**Reply by aserowy** [reply, marked as answer]
Seems fixed in the upcoming version i guess. Running the given snippets, i can reproduce the fail in the shell directly.

**Reply by aserowy** [reply]
tyvm @fdncred

**Reply by fdncred** [reply]
feel free to build from the repo and confirm it's fixed. after i clone the repo, i usually cd to the nushell folder and do `cargo install --path . --features=dataframe,extra,mimalloc`
